qiskit_nature/operators/second_quantization/spin_op.py

```diff
@@ -12,117 +12,479 @@
+
+        SpinOp(
+            [
+                ("XX", -1),
+                ("YY", -1),
+                ("ZZ", -1),
+                ("ZI", -0.3),
+                ("IZ", -0.3),
+            ],
+            spin=1
+        )
+
+    This means :math:`- X_1 X_0 - Y_1 Y_0 - Z_1 Z_0 - 0.3 Z_0 - 0.3 Z_1`.
+
+    :class:`SpinOp` can be initialized with internal data structure (`numpy.ndarray`) directly.
+    In this case, `data` is a tuple of two elements: `spin_array` and `coeffs`.
+    `spin_array` is 3-dimensional `ndarray`. 1st axis has three elements 0, 1, and 2 corresponding
+    to x, y, and z.  2nd axis represents the index of terms.
+    3rd axis represents the index of register.
+    `coeffs` is one-dimensional `ndarray` with the length of the number of terms.
+
+    **Algebra**
+
+    :class:`SpinOp` supports the following basic arithmetic operations: addition, subtraction,
+    scalar multiplication, and dagger(adjoint).
+    For example,
+
+    .. jupyter-execute::
+
+        from qiskit_nature.operators import SpinOp
+
+        x = SpinOp("X", spin=3/2)
+        y = SpinOp("Y", spin=3/2)
+        z = SpinOp("Z", spin=3/2)
+
+        print("Raising operator:")
+        print(x + 1j * y)
+        plus = SpinOp("+", spin=3/2)
+        print("This is same with: ", plus)
+        print("Lowering operator:")
+        print(x - 1j * y)
+        minus = SpinOp("-", spin=3/2)
+        print("This is same with: ", minus)
+
+        print("Dagger")
+        print(~(1j * z))
+
     """
-    Spin type operators. This class represents sums of `spin strings`, i.e. linear combinations of
-    SpinOperators with same spin and register length.
-    """
 
-    def __init__(self, operator_list):
-        # 1. Parse input
-        self._register_length = len(operator_list[0])
-        self.spin = operator_list[0].spin
-        for elem in operator_list:
-            assert isinstance(elem, SpinOperator)
-            assert len(elem) == self._register_length, \
-                'Cannot sum operators acting on registers of different length'
-            assert elem.spin == self.spin, \
-                'Cannot sum operators with different spins.'
+    _XYZ_DICT = {"X": 0, "Y": 1, "Z": 2}
+    _VALID_LABEL_PATTERN = re.compile(
+        r"^([IXYZ\+\-]_\d(\^\d)?\s)*[IXYZ\+\-]_\d(\^\d)?(?!\s)$|^[IXYZ\+\-]+$"
+    )
+    _SPARSE_LABEL_PATTERN = re.compile(r"^([IXYZ]_\d(\^\d)?\s)*[IXYZ]_\d(\^\d)?(?!\s)$")
+    _DENSE_LABEL_PATTERN = re.compile(r"^[IXYZ]+$")
+
+    def __init__(
+        self,
+        data: Union[
+            str,
+            List[Tuple[str, complex]],
+            Tuple[np.ndarray, np.ndarray],
+        ],
+        spin: Union[float, Fraction] = Fraction(1, 2),
+    ):
+        r"""``SpinOp``.
+
+        Args:
+            data: label string or list of labels and coefficients. See documentation of
+                  :class:`SpinOp` for more details.
+            spin: positive half-integer (integer or half-odd-integer) that represents spin.
```

Comment:
```suggestion
            spin: positive half-integral or integral value.
```

Reply:
https://en.wikipedia.org/wiki/Half-integer

Reply:
I know that technically "half-integer" is correct. But I agree with @woodsp-ibm that it may be confusing to read about integers while the code does not accept `int`. If you feel strongly about leaving "half-integer" written as such, please make it a link to exactly this wikipedia entry as this should clear up any confusion for the reader.

Reply:
`int` is also acceptable.

Reply:
Yes, `int` is a valid type to give when its says `float` for a typehint. Supporting `Fraction` seems fine too as a better way to pass in a 1/2 etc.